contracts/MarginCalculator.sol

```diff
@@ -43,19 +45,288 @@ contract MarginCalculator {
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+        // get array of expiries
+        uint256[] storage expiryArray = productTimeToExpiry[productHash];
+
+        // check that this is the first expiry to set, if not the last expiry should be less than the new one to insert (to make sure the array stay in order)
+        require(
+            (expiryArray.length == 0) || (_timeToExpiry > expiryArray[expiryArray.length.sub(1)]),
+            "MarginCalculator: expiry array is not in order"
+        );
+
+        // check that this time to expiry have an upper bound value (setTimeToExpiryValue() should be called before this)
+        require(
+            timeToExpiryValue[productHash][_timeToExpiry] != 0,
+            "MarginCalculator: no expiry upper bound value found"
+        );
+
+        // add new time to expiry to array
+        expiryArray.push(_timeToExpiry);
+    }
+
+    /**
+     * @notice set option upper bound value for specific time to expiry (1e27)
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     * @param _value upper bound value
+     */
+    function setTimeToExpiryValue(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry,
+        uint256 _value
+    ) external onlyOwner {
+        // check that upper bound value is not zero
+        require(_value > 0, "MarginCalculator: invalid option upper bound value");
+
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+
+        // add new upper bound value for this product at specific time to expiry
+        timeToExpiryValue[productHash][_timeToExpiry] = _value;
+    }
+
+    /**
+     * @notice set spot shock value scaled to 1e27
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _shockValue spot shock value
+     */
+    function setSpotShock(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _shockValue
+    ) external onlyOwner {
+        require(_shockValue > 0, "MarginCalculator: invalid spot shock value");
+
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+
+        // set spot shock value in 27 decimals
+        spotShock[productHash] = _shockValue;
+    }
+
+    /**
+     * @notice set oracle deviation (1e27)
+     * @dev can only be called by owner
+     * @param _deviation deviation value
+     */
+    function setOracleDeviation(uint256 _deviation) external onlyOwner {
```

Comment:
Might make sense this should be a mapping per product as chainlink deviations are different between products.  COMPUSD is 1% ETHUSD 0.5%, LINKUSD 1%, etc.

We could also just be conservative and use 2% or something.

Reply:
What do u think is better ? if keeping it conservative will work, I prefer that rather than updating now.